src/main/kotlin/dev/slne/surf/database/redis/RedisProvider.kt

```diff
@@ -0,0 +1,161 @@
+package dev.slne.surf.database.redis
+
+import dev.slne.surf.database.config.ConnectionConfig
+import dev.slne.surf.database.redis.listener.RedisEventHandler
+import dev.slne.surf.database.redis.listener.RedisPacketEvent
+import dev.slne.surf.database.redis.packet.RedisPacket
+import dev.slne.surf.database.serializer.SurfSerializer
+import dev.slne.surf.database.utils.callMethodWithRedisPacketEvent
+import dev.slne.surf.database.utils.getAnnotatedMethods
+import io.lettuce.core.*
+import io.lettuce.core.api.StatefulRedisConnection
+import io.lettuce.core.api.coroutines
+import io.lettuce.core.api.coroutines.RedisStringCoroutinesCommands
+import io.lettuce.core.pubsub.StatefulRedisPubSubConnection
+import io.lettuce.core.pubsub.api.async.RedisPubSubAsyncCommands
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.PolymorphicSerializer
+import kotlin.reflect.KFunction
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+import kotlin.time.toJavaDuration
+
+@OptIn(ExperimentalLettuceCoroutinesApi::class)
+class RedisProvider(private val connectionConfig: ConnectionConfig) {
+
+    private var client: RedisClient? = null
+    private var connection: StatefulRedisConnection<String, String>? = null
+    private var pubSubConnection: StatefulRedisPubSubConnection<String, String>? = null
+    private var _commands: RedisStringCoroutinesCommands<String, String>? = null
+    private var _pubSub: RedisPubSubAsyncCommands<String, String>? = null
+
+    val commands get() = _commands
+    val pubSub get() = _pubSub
+
+    private var _scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
```

Comment:
```suggestion
    private var _scope = CoroutineScope(Dispatchers.IO + SupervisorJob() + CoroutineName("redis-connection-scope-{identifier}"))
```
for easier debugging and identifications in stacktraces. Maybe you also want to add an [CoroutineExceptionHandler](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-exception-handler/) -- https://kotlinlang.org/docs/exception-handling.html